modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ExpressionLanguageTransformationValueResolver.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.resolver;
+
+import static org.mule.runtime.api.metadata.DataType.fromObject;
+import static org.mule.runtime.api.metadata.DataType.fromType;
+import static org.mule.runtime.core.api.util.ClassUtils.isInstance;
+
+import org.mule.runtime.api.el.BindingContext;
+import org.mule.runtime.api.el.ExpressionLanguage;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.api.metadata.DataType;
+import org.mule.runtime.api.metadata.TypedValue;
+
+/**
+ * Value resolver that relies on a expression language transformation to convert the delegate resolver into the expected type.
+ * 
+ * @since 4.5
+ */
+public class ExpressionLanguageTransformationValueResolver implements ValueResolver {
+
+  private static final String PAYLOAD_IDENTIFIER = "payload";
+  private static final String PAYLOAD_EXPRESSION = "#[" + PAYLOAD_IDENTIFIER + "]";
+
+  private final Class expectedType;
+  private final ValueResolver valueResolverDelegate;
+  private final ExpressionLanguage expressionLanguage;
+
+  public ExpressionLanguageTransformationValueResolver(ValueResolver valueResolverDelegate, Class expectedType,
+                                                       ExpressionLanguage expressionLanguage) {
+    this.expectedType = expectedType == null ? Object.class : expectedType;
+    this.valueResolverDelegate = valueResolverDelegate;
+    this.expressionLanguage = expressionLanguage;
+  }
+
+  @Override
+  public Object resolve(ValueResolvingContext context) throws MuleException {
+    Object resolvedValue = valueResolverDelegate.resolve(context);
+    DataType dataType;
+    if (resolvedValue instanceof TypedValue) {
+      dataType = ((TypedValue<?>) resolvedValue).getDataType();
+      resolvedValue = ((TypedValue<?>) resolvedValue).getValue();
+    } else {
+      dataType = fromObject(resolvedValue);
+    }
+    if (isInstance(expectedType, resolvedValue)) {
+      return resolvedValue;
```

Comment:
bug. If the resolvedValue vas a TypedValue, you need to return the original TypedValue, not the unwrapped one